{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/miniconda3/envs/gcloud/lib/python3.13/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from synthesizeme.personalrm.synthesizeme import SynthesizeMe\n",
    "from synthesizeme.utils.convert import convert_wildfeedback_csv_to_json"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Collect User Feedback"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing user: 95dd8505-37e0-43c7-af9f-a185d154d8c9 with 14 entries\n"
     ]
    }
   ],
   "source": [
    "data = convert_wildfeedback_csv_to_json('../conversations/user0.csv')['95dd8505-37e0-43c7-af9f-a185d154d8c9']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## View the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'context': [{'role': 'user',\n",
       "   'content': 'Can you summarize the methods that are outlined here - include high level steps, some useful details (can exclude irrelevant concentrations/percentages) and then finish with the high level chronological summary of all the steps? Methods\\nCell culture\\nHEK 293T, 3T3, L929, Raw 264.7, DLD-1, A549 and MCF7 cells were all obtained from ATCC and were maintained in DMEM supplemented with 10% fetal bovine serum (FBS; vol/vol), 100\\u2009U\\u2009ml–1 penicillin, 100\\u2009μg\\u2009ml–1 streptomycin and 1× GlutaMAX.\\n\\nIMR90 fibroblasts were obtained from ATCC (CCL-186) and maintained in EMEM (ATCC, 30-2003) supplemented with 10% (vol/vol) FBS.\\n\\nFor iNeurons, iPS cells (ALSTEM, iP11N) were differentiated using a strategy combining NGN2 programming and small-molecule patterning optimized from a previously published protocol68. On day 7 of differentiation, cells were lifted from the flask and plated at 40,000 cells per well in poly-d-lysine- and iMatrix-coated 96-well plates. On day 8, cells were transduced by adding appropriate volumes of lentiviral supernatant overnight to achieve a multiplicity of infection (MOI) of <0.3. Cells were then subjected to antibiotic selection by puromycin (Gibco, A1113803) at 1\\u2009μg\\u2009ml–1 on day 11 for 3\\u2009days. At day 18, cells were fixed for 30\\u2009min in 4% PFA in 1× PBS for ISS.\\n\\niAstrocytes were generated from iPS cells (ALSTEM, iP11) based on a published protocol69. Immature astrocytes collected from astrospheres were plated at 10,000 cells per well in Matrigel-coated 96-well plates for further maturation. A day later, cells were transduced by adding appropriate volumes of lentiviral supernatant overnight to achieve an MOI of <0.3. Cells were selected by puromycin treatment at 2\\u2009μg\\u2009ml–1 at day 5 of maturation for 3\\u2009days and fixed at day 8 with 4% PFA in 1× PBS for 30\\u2009min.\\n\\nImmortalized macrophages (ER-Hoxb8-immortalized mouse myeloid progenitor cells70,71) were established from Cas9 transgenic mice72. Cells were cultured in RPMI 1640 medium supplemented with 10% (vol/vol) FBS, 20\\u2009ng\\u2009ml–1 mouse granulocyte–macrophage colony-stimulating factor (eBioscience) and 1\\u2009mM β-estradiol (Millipore Sigma). Cells were transduced by spinfection at 3,200g at 32\\u2009°C for 45\\u2009min with 5\\u2009μg\\u2009ml–1 polybrene, cultured for another 2\\u2009days and selected with 10\\u2009μg\\u2009ml–1 puromycin. Cells were differentiated into macrophages in DMEM supplemented with 10% (vol/vol) FBS and 20% (vol/vol) L929-conditioned medium at 37\\u2009°C with 5% CO2 and were replated in 96-well plates at 20,000 cells per well on day 5 for experiments.\\n\\nBone marrow was collected from tibias and femurs of 12- to 18-week-old Cas9–enhanced green fluorescent protein (eGFP) transgenic mice73 or 6- to 12-week-old Cas9 transgenic mice. After red blood cells lysis (ACK lysis buffer, Gibco, A1049201), 20 million cells were plated in BMDM medium (DMEM high glucose, 10% Tet-Negative heat-inactivated FBS, 1× GlutaMAX, 100\\u2009U\\u2009ml–1 penicillin–streptomycin and 100\\u2009ng\\u2009ml–1 recombinant mouse macrophage colony-stimulating factor (Genentech media facility)) at a density of 0.8 million cells per ml in 150-mm non-tissue culture-treated dishes (Corning). Cells were transduced on day 3 by adding a single lentivirus or lentiviral pool supernatants overnight in fresh BMDM medium. Transduced cells were selected with 5\\u2009μg\\u2009ml–1 puromycin from days 8 to 11.\\n\\nT cells were isolated using a pan-T cell isolation kit (Miltenyi) and frozen in FBS supplemented with 10% DMSO. Following thawing, T cells were resuspended in T cell medium (X VIVO 15\\u2009+\\u20095% FBS\\u2009+\\u2009500\\u2009U\\u2009ml–1 IL-2) to a concentration of 1\\u2009×\\u2009106\\u2009cells per ml and stimulated with anti-human CD3/CD28 CTS Dynabeads (Thermo Fisher Scientific, 40203D) at a 1:1 cell:bead ratio. Approximately 18\\u2009h after stimulation, cells were transduced with Lenti-Cas9 (2.5% vol/vol). Forty-two hours after stimulation, cells were transferred to a 96-well plate and transduced with CD81 sgRNA virus at an MOI of <0.2. Cells were then selected with 2\\u2009µg\\u2009ml–1 puromycin on day 3. Over the next few days, cells were expanded and maintained by adding medium or splitting as necessary. On day 6, CRISPR-mediated Cd81 knockout was assessed by flow cytometry. Cells were then plated onto anti-CD3-coated lysine surfaces and allowed to adhere for 15\\u2009min at 37\\u2009°C before 4% PFA fixation and ISS.\\n\\nAll cells were cultured at 37\\u2009°C and 5% CO2 in a humidified incubator.\\n\\nAnimal experiments\\nAll protocols involving animals were approved by Genentech’s Institutional Animal Care and Use Committee, in accordance with guidelines that adhere to and exceed state and national ethical regulations for animal care and use in research. All mice were maintained in a pathogen-free animal facility under standard animal room conditions (temperature of 21\\u2009±\\u20091\\u2009°C, humidity between 55 and 60% and 12-h light/12-h dark cycle).\\n\\nImaging\\nImaging was performed on a Nikon Ti2 microscope equipped with Nikon lenses, a Hamamatsu Fusion BT camera or Teledyne Photometrics Kinetix sCMOS and a Lumencor CELESTA light engine.\\n\\nISS cycles were imaged using either a ×10/0.45-NA CFI Plan Apo Lambda objective or a ×4/0.20-NA CFI Plan Apo Lambda objective with the CELESTA-DA/FI/TR/Cy5/Cy7 (Semrock) excitation filter and FF01-391/477/549/639/741 dichroic filter unless noted otherwise. Laser excitation wavelength and emission filters were set up for each channel: DAPI (408\\u2009nm, FF01-441/511/593/684/817), base G (545\\u2009nm, FF01-575/19), base T (545\\u2009nm, FF01-615/24), base A (637\\u2009nm, FF01-680/42) and base C (637\\u2009nm, FF01-732/68).\\n\\nFor p65 phenotyping, the following conditions were used for imaging immunofluorescence signal with a ×10 objective: DAPI (365\\u2009nm, FF01-433/24) and CF430 (446\\u2009nm, 5%, excitation FF01-432/523/702, emission FF01-470/28, dichroic mirror FF459/526/596-Di01).\\n\\nHCR FISH phenotypic images were acquired with a ×20/0.75-NA CFI Plan Apo Lambda objective with the following conditions: DAPI (365\\u2009nm, FF01-441/511/593/684/817), AF594 (561\\u2009nm, FF01-615/24), AF647 (637\\u2009nm, FF01-680/42) and AF750 (748\\u2009nm, FF01-792/64).\\n\\nIBEX immunofluorescence images were acquired with a ×20/0.75-NA CFI Plan Apo Lambda objective with the following conditions: DAPI (365\\u2009nm, FF01-441/511/593/684/817), CF430 (446\\u2009nm, excitation 432/523/702, emission FF01-470/28, dm 459/526/596), GFP (477\\u2009nm, FF01-511/20), AF532 (545\\u2009nm, excitation FF01-432/523/702, emission FF01-563/9, dm 555-Di03), AF647 (637\\u2009nm, FF01-680/42) and AF750 (748\\u2009nm, emission FF01-792/64).\\n\\nFor imaging time comparisons between the conventional protocol with ×10 magnification and PerturbView with ×4 magnification, the following exposures were used: DAPI (200\\u2009ms), base G (200\\u2009ms), base C (200\\u2009ms), base A (200\\u2009ms) and base T (800\\u2009ms). Ten percent and 4% overlapping regions were acquired for ×10 and ×4 objective lenses, respectively, to use the same physical distance for the image registration. The same exposure was also used for the frameshift reporter assay at ×4 magnification.\\n\\nLentiviral transduction\\nHEK293T cells were seeded into six-well plates at a density of 1 million cells per well. The following day, cells were transfected with the expression plasmid, delta8.9 and pCMV-VSV-G at a molar ratio of (1:2.3:0.2) using Lipofectamine 3000 (Thermo Fisher Scientific, L3000015). Opti-MEM (Gibco, 31985062) was exchanged to culture medium after 4\\u2009h. Viral supernatant was collected 2\\u2009days after transfection and filtered through a 0.45-μm filter. For infection of BMDMs, viral supernatant was further concentrated by adding Lenti-X concentrator (Takara, 31231) and centrifuging at 4\\u2009°C for 1\\u2009h. Lentiviral titer was quantified before the screens using a CellTiter-Glo kit (Promega, G7570). For all experiments, an MOI of <0.2 was used, except in the cases of iAstrocytes and iNeurons (MOI\\u2009<\\u20090.3).\\n\\nISS\\nCells were fixed with 4% PFA (Electron Microscopy Sciences, 15714) in PBS for 30\\u2009min and washed three times with PBS with 0.05% Tween (PBST). Cells were permeabilized with 70% ethanol for 30\\u2009min, followed by 75% volume exchange with PBST three times to prevent drying and two complete PBST washes. Afterward, heat decross-linking was performed in 0.1\\u2009M sodium bicarbonate and 0.3\\u2009M NaCl in water at 65\\u2009°C for 4\\u2009h (cell culture) to 24\\u2009h (tissue). It is important to use glass-bottom plates that tolerate high heat (Cellviz, P96-1.5H-N or P12-1.5H-N). After three washes with PBST, IVT was conducted using a HiScribe T7 Quick High Yield RNA Synthesis kit (New England Biolabs, E2050S), 1× T7 RNA Polymerase Mix, NTP Buffer Mix at 10\\u2009mM and the addition of RiboLock RNase inhibitor (Thermo Fisher Scientific, EO0384) at 0.4\\u2009U\\u2009μl–1 at 37\\u2009°C for 12\\u2009h. For control samples without IVT, fixed cells were maintained in PBST with 0.4\\u2009U\\u2009μl–1 Ribolock at 4\\u2009°C for less than 48\\u2009h after permeabilization.\\n\\nA primer cross-linking protocol was adopted from Labitigan et al.21 for all data (referred to as conventional OPS) except those presented in Fig. 1c, which use the original OPS protocol from Feldman et al.13. An amine-modified primer was hybridized at 1\\u2009μM in PBST for 30\\u2009min at room temperature. After the hybridization and before reverse transcription, cells were fixed with 3.2% PFA and 0.1% glutaraldehyde (Electron Microscopy Sciences, 16120) in PBST for 30\\u2009min. The reaction was quenched by adding 0.2\\u2009M Tris-HCl (pH 8), and cells were washed three times with PBST. The reverse transcription reaction mix (1× RevertAid RT buffer, 250\\u2009μM dNTPs (New England Biolabs, N0447L), 0.2\\u2009mg\\u2009ml–1 recombinant albumin (New England Biolabs, B9200S), 1\\u2009μM reverse transcription primer, 0.4\\u2009U\\u2009μl–1 Ribolock RNase inhibitor and 4.8\\u2009U\\u2009μl–1 RevertAid H minus reverse transcriptase (Thermo Fisher Scientific, EP0452)) was added to the samples, followed by incubation at 37\\u2009°C overnight.\\n\\nAfter reverse transcription, cells were washed three times with PBST and fixed again in 3.2% PFA and 0.1% glutaraldehyde in PBST for 30\\u2009min. Cells were then washed with PBST six times to remove residual dNTPs. After this step, cells were incubated in gap-fill reaction mix (1× Ampligase buffer, 0.4\\u2009U\\u2009μl–1 RNase H (Enzymatics, Y9220L), 0.2\\u2009mg\\u2009ml–1 recombinant albumin, 100\\u2009nM padlock probe, 0.02\\u2009U\\u2009μl–1 TaqIT polymerase (Enzymatics, P7620L), 0.5\\u2009U\\u2009μl–1 Ampligase (Lucigen, A3210K) and 50\\u2009nM dNTPs) for 5\\u2009min at 37\\u2009°C and 90\\u2009min at 45\\u2009°C. Cells were washed with PBST three times, and rolling circle amplification was performed by adding the reaction mix (1× Phi29 buffer, 250\\u2009μM dNTPs, 0.2\\u2009mg\\u2009ml–1 recombinant albumin, 5% glycerol and 1\\u2009U\\u2009μl–1 Phi29 DNA polymerase (Thermo Fisher Scientific, EP0091)) and incubating at 30\\u2009°C for >20\\u2009h.\\n\\nISS steps were conducted according to the OPS protocol30,13 with the primer sequences listed in Supplementary Table 6. See the Supplementary Note for the detailed protocol.\\n\\nChimeric promoter designs\\nThe nucleosome binding sequence of the U6 promoter was modified to incorporate the T7 promoter while keeping the downstream proximal sequence element and upstream distal sequence element intact34. Given the uncertain tradeoff between U6 and T7 promoter sequences, we adopted multiple strategies for embedding the T7 promoter into the U6 promoter. The modification included the integration of the minimal T7 promoter sequence (TAATACGACTCACTATAG) or supplemented with trailing (GAATT) and/or leading (AGAGT or ATAAT) sequences to enhance transcriptional activity31,74. One construct lacked the initial T base of the promoter, and three constructs were truncated as per the configurations described in the published Mini-U6 promoter34. In most constructs, we strategically placed modifications before the existing G or GG sequences to fulfill the known sequence constraints (G at positions +1 to +3 for efficient transcription) while minimizing modifications31,74.\\n\\nAssessment of editing efficiency\\nFor determining editing efficiency, a targeting sgRNA was transduced by lentivirus to Cas9-expressing cells. Next, cells were cultured for 4\\u2009days and fixed in 4% PFA in PBS for 30\\u2009min. Cells were washed twice with PBS and incubated with fluorophore-conjugated antibodies (20× dilution in PBS; anti-CD81 (BioLegend, 349501) or anti-CD44 (Biolegend, 103023)) for 1\\u2009h, followed by a PBS wash. Antibody signal was measured by flow cytometry (Sony, MA900 Multi-Application Cell Sorter) gated on live cells based on forward and side scatter.\\n\\nPerturbView signal quantification\\nCell models were fixed in 4% PFA. After 30\\u2009min, cells were washed with PBST three times. Plates without PerturbView were reserved at 4\\u2009°C in PBS\\u2009+\\u20090.4\\u2009U\\u2009μl–1 Ribolock while decross-linking, and IVT protocols were performed. Ethanol permeabilization and reverse transcription were performed within 72\\u2009h after fixation.\\n\\nBecause the massive signal amplification of PerturbView poses a challenge in fitting read intensities within the optics’ dynamic range when using the same exposure setting with the standard OPS protocol, a fourfold longer duration of exposure was used for non-PerturbView images of iAstrocytes, fibroblasts and iNeurons.\\n\\nFrameshift reporter assay\\nCas9-expressing A549 cells were first transduced with the frameshift reporter13 at a high MOI by repeating the transduction processes and selected with 300\\u2009μg\\u2009ml–1 hygromycin (Thermo Fisher Scientific, 10687010) for 1 week. Five targeting and five nontargeting control sgRNAs were individually cloned into a PerturbView vector, pooled equimolarly, packaged as lentivirus and transduced into Cas9-expressing A549 cells at a low MOI, followed by puromycin selection. Cells were plated at 12,000 cells per well in a 96-well plate (CellViz, P96-1.5H-N) and fixed with 4% PFA for 30\\u2009min the following day. Epitope staining was conducted as previously described13, followed by ISS. As described above, fourfold longer durations of exposure were used for standard OPS than for PerturbView due to the different dynamic ranges of the two methods.\\n\\nFor comparison of the fixation methods, cells were fixed with 4% PFA for 30\\u2009min or with a 3:1 mix of methanol and acetic acid for 20\\u2009min. For the decross-linking step, cells were incubated in the decross-linking buffer either at room temperature (25\\u2009°C) or at 65\\u2009°C for 4\\u2009h.\\n\\nLibrary construction for the BMDM screen\\nFor the BMDM library, four sgRNAs were designed per gene for 23 NF-κB-related genes13, 130 metabolic and signaling genes (Supplementary Table 2) and 10 olfactory receptor genes (as nonexpressed gene controls), along with 60 different nontargeting sgRNAs. Spacer sequences were generated by crisprVerse75 to design barcodes that can be uniquely identified by six and eight cycles of ISS for NTC and BMDM libraries, respectively. The oligonucleotide pool was synthesized by Twist Bioscience, amplified by dial-out PCR76 using Ex Premier DNA Polymerase (Takara Bio, RR370A), cloned into a PerturbView vector with Golden Gate cloning using BsmBI (New England Biolabs, E1602S), transformed into Stellar electrocompetent cells (Takara Bio, 636765) and packaged into a lentivirus library while maintaining >1,000× coverage at every step.\\n\\nTo validate sgRNA transduction, genomic DNA was extracted and amplified according to an established protocol30. Indexed amplicons were sequenced on an Illumina MiSeq with 150 cycles and a v3 reagent kit (Illumina, MS-102-3001). The reads were aligned to the expected amplicon sequences (N20 corresponding to the spacer region flanked by backbone sequences) using the Burrows–Wheeler alignment tool (bwa-mem)77 to produce read counts for each designed guide.\\n\\nNF-κB translocation screens\\nBMDMs were isolated from mice expressing Cas9–eGFP and cultured in DMEM high glucose, 10% FBS, 2\\u2009mM glutamine, 1% penicillin–streptomycin and 100\\u2009ng\\u2009ml–1 colony-stimulating factor 1. After 2\\u2009days, these cells were split into new plates and infected with the guide-bearing lentiviruses at an MOI of <0.2. The medium was replaced the following day, and selection of the infected BMDMs began with 5\\u2009µg\\u2009ml–1 puromycin. Three days after the selection and medium change, cells were plated in a 12-well plate (Cellviz, P12-1.5H-N) at a density of 276,000 cells per well. After 24\\u2009h, cells were stimulated with 10\\u2009ng\\u2009ml–1 TNF (Biolegend, 575202), 2\\u2009ng\\u2009ml–1 IL-1β (Biolegend, 575108), 100\\u2009ng\\u2009ml–1 LPS (Enzo, ALX-581-013-L001) and water for 40\\u2009min before fixation. Cells were fixed in 4% PFA in PBS for 30\\u2009min. Following three PBST washes, cells were permeabilized with 0.2% Triton X-100 (Sigma-Aldrich, 93443) in PBS for 10\\u2009min. Cells were then blocked with 3% bovine serum albumin (BSA) in PBS for 45\\u2009min, incubated with anti-p65 in PBS at 1:1,000 (Cell Signaling Technology, 8242) for 1\\u2009h and washed three times with PBST. The medium was replaced with donkey anti-rabbit CF430 IgG (Biotium, 20460) diluted 1:500 in 3% BSA in PBS and incubated for 30\\u2009min, followed by three washes with PBST. Before imaging, the medium was replaced with 200\\u2009ng\\u2009ml–1 DAPI (Sigma-Aldrich, D9542) in 2× saline-sodium citrate (SSC; Invitrogen, 15557044) buffer for 10\\u2009min. After antibody imaging, ISS was performed beginning from the ethanol permeabilization step.'}],\n",
       " 'chosen': {'role': 'assistant',\n",
       "  'content': \"I'll break down the methods outlined in the text into high-level steps, useful details, and a chronological summary.\\n\\n**Cell Culture Methods**\\n\\n1. **Cell Line Maintenance**: Various cell lines (HEK 293T, 3T3, L929, Raw 264.7, DLD-1, A549, and MCF7) were obtained from ATCC and maintained in DMEM supplemented with 10% FBS, 100 U/ml penicillin, 100 μg/ml streptomycin, and 1× GlutaMAX.\\n2. **iNeurons Generation**: iPS cells (ALSTEM, iP11N) were differentiated using a strategy combining NGN2 programming and small-molecule patterning. Cells were plated at 40,000 cells per well in poly-d-lysine- and iMatrix-coated 96-well plates on day 7 of differentiation.\\n3. **iAstrocytes Generation**: iPS cells (ALSTEM, iP11) were differentiated into immature astrocytes, which were then plated at 10,000 cells per well in Matrigel-coated 96-well plates for further maturation.\\n4. **Immortalized Macrophages Generation**: ER-Hoxb8-immortalized mouse myeloid progenitor cells were established from Cas9 transgenic mice and cultured in RPMI 1640 medium supplemented with 10% FBS, 20 ng/ml mouse granulocyte–macrophage colony-stimulating factor, and 1 mM β-estradiol.\\n5. **Bone Marrow-Derived Macrophages (BMDMs) Generation**: Bone marrow was collected from tibias and femurs of Cas9–eGFP transgenic mice and plated in BMDM medium. Cells were transduced with lentivirus on day 3 and selected with puromycin from days 8 to 11.\\n6. **T Cell Isolation**: T cells were isolated using a pan-T cell isolation kit and frozen in FBS supplemented with 10% DMSO. Following thawing, T cells were resuspended in T cell medium and stimulated with anti-human CD3/CD28 CTS Dynabeads.\\n\\n**Lentiviral Transduction**\\n\\n1. **HEK 293T Cell Transfection**: HEK 293T cells were transfected with the expression plasmid, delta8.9, and pCMV-VSV-G using Lipofectamine 3000.\\n2. **Viral Supernatant Collection**: Viral supernatant was collected 2 days after transfection and filtered through a 0.45-μm filter.\\n3. **Lentiviral Titer Quantification**: Lentiviral titer was quantified using a CellTiter-Glo kit.\\n4. **Cell Infection**: Cells were infected with lentivirus at an MOI of <0.2, except for iAstrocytes and iNeurons, which were infected at an MOI of <0.3.\\n\\n**ISS (In Situ Sequencing) Methods**\\n\\n1. **Cell Fixation**: Cells were fixed with 4% PFA in PBS for 30 min and washed three times with PBS with 0.05% Tween.\\n2. **Heat Decross-linking**: Cells were incubated in 0.1 M sodium bicarbonate and 0.3 M NaCl in water at 65°C for 4-24 h.\\n3. **IVT (In Vitro Transcription)**: IVT was conducted using a HiScribe T7 Quick High Yield RNA Synthesis kit.\\n4. **Primer Cross-linking**: An amine-modified primer was hybridized at 1 μM in PBST for 30 min at room temperature.\\n5. **Reverse Transcription**: The reverse transcription reaction mix was added to the samples, followed by incubation at 37°C overnight.\\n6. **Gap-Fill Reaction**: Cells were incubated in gap-fill reaction mix for 5 min at 37°C and 90 min at 45°C.\\n7. **Rolling Circle Amplification**: Rolling circle amplification was performed by adding the reaction mix and incubating at 30°C for >20 h.\\n\\n**Assessment of Editing Efficiency**\\n\\n1. **Targeting sgRNA Transduction**: A targeting sgRNA was transduced by lentivirus to Cas9-expressing cells.\\n2. **Cell Culture**: Cells were cultured for 4 days and fixed in 4% PFA in PBS for 30 min.\\n3. **Flow Cytometry**: Cells were incubated with fluorophore-conjugated antibodies and measured by flow cytometry.\\n\\n**PerturbView Signal Quantification**\\n\\n1. **Cell Fixation**: Cells were fixed in 4% PFA for 30 min.\\n2. **Ethanol Permeabilization**: Cells were permeabilized with 70% ethanol for 30 min.\\n3. **Reverse Transcription**: The reverse transcription reaction mix was added to the samples, followed by incubation at 37°C overnight.\\n4. **ISS**: ISS was performed beginning from the ethanol permeabilization step.\\n\\n**Frameshift Reporter Assay**\\n\\n1. **Cas9-Expressing A549 Cell Transduction**: Cas9-expressing A549 cells were transduced with the frameshift reporter at a high MOI.\\n2. **PerturbView Vector Transduction**: Cells were transduced with PerturbView vector at a low MOI.\\n3. **Puromycin Selection**: Cells were selected with 300 μg/ml hygromycin for 1 week.\\n4. **ISS**: ISS was performed after fixation.\\n\\n**NF-κB Translocation Screens**\\n\\n1. **BMDM Isolation**: BMDMs were isolated from mice expressing Cas9–eGFP.\\n2. **Lentivirus Infection**: BMDMs were infected with guide-bearing lentiviruses at an MOI of <0.2.\\n3. **Puromycin Selection**: Selection of the infected BMDMs began with 5 μg/ml puromycin.\\n4. **TNF, IL-1β, and LPS Stimulation**: Cells were stimulated with TNF, IL-1β, and LPS for 40 min.\\n5. **Fixation**: Cells were fixed in 4% PFA in PBS for 30 min.\\n6. **p65 Antibody Imaging**: Cells were incubated with anti-p65 antibody and measured by flow cytometry.\\n7. **ISS**: ISS was performed beginning from the ethanol permeabilization step.\\n\\n**Chronological Summary**\\n\\n1. Cell line maintenance and generation of iNeurons, iAstrocytes, immortalized macrophages, and BMDMs.\\n2. Lentiviral transduction of cells with targeting sgRNA.\\n3. Assessment of editing efficiency by flow cytometry.\\n4. PerturbView signal quantification by ISS.\\n5. Frameshift reporter assay by ISS.\\n6. NF-κB translocation screens by ISS.\\n\\nNote: This summary is a high-level overview of the methods outlined in the text. The actual protocols and procedures may vary depending on the specific experimental design and requirements.\"},\n",
       " 'rejected': {'role': 'assistant',\n",
       "  'content': \"I'm so excited to dive into this complex protocol and break it down for you. There are several cell types and methods involved, but I'll try to summarize the high-level steps and details for each.\\n\\n**Cell Culture**\\n\\n* Various cell lines (HEK 293T, 3T3, L929, Raw 264.7, DLD-1, A549, and MCF7) were obtained from ATCC and maintained in DMEM supplemented with 10% FBS, penicillin, streptomycin, and GlutaMAX.\\n* IMR90 fibroblasts were obtained from ATCC and maintained in EMEM supplemented with 10% FBS.\\n* iNeurons were generated from iPS cells using a strategy combining NGN2 programming and small-molecule patterning.\\n* iAstrocytes were generated from iPS cells based on a published protocol.\\n* Immortalized macrophages (ER-Hoxb8-immortalized mouse myeloid progenitor cells) were established from Cas9 transgenic mice.\\n* Bone marrow-derived macrophages (BMDMs) were isolated from Cas9–eGFP transgenic mice.\\n* T cells were isolated using a pan-T cell isolation kit and frozen in FBS supplemented with 10% DMSO.\\n\\n**Lentiviral Transduction**\\n\\n* HEK 293T cells were transfected with expression plasmids, delta8.9, and pCMV-VSV-G using Lipofectamine 3000.\\n* Viral supernatant was collected 2 days after transfection and filtered through a 0.45-μm filter.\\n* Lentiviral titer was quantified using a CellTiter-Glo kit.\\n* Cells were transduced with lentiviral supernatant at an MOI of <0.2, except for iAstrocytes and iNeurons (MOI < 0.3).\\n\\n**ISS (In Situ Sequencing)**\\n\\n* Cells were fixed with 4% PFA in PBS for 30 min and washed three times with PBST.\\n* Cells were permeabilized with 70% ethanol for 30 min, followed by 75% volume exchange with PBST three times.\\n* Heat decross-linking was performed in 0.1 M sodium bicarbonate and 0.3 M NaCl in water at 65°C for 4-24 h.\\n* IVT (In Vitro Transcription) was conducted using a HiScribe T7 Quick High Yield RNA Synthesis kit.\\n* Primer cross-linking protocol was adopted from Labitigan et al. (referred to as conventional OPS) or Feldman et al. (original OPS protocol).\\n* Reverse transcription reaction mix was added to the samples, followed by incubation at 37°C overnight.\\n* Gap-fill reaction mix was incubated in the samples for 5 min at 37°C and 90 min at 45°C.\\n* Rolling circle amplification was performed by adding the reaction mix and incubating at 30°C for >20 h.\\n\\n**Assessment of Editing Efficiency**\\n\\n* A targeting sgRNA was transduced by lentivirus to Cas9-expressing cells.\\n* Cells were cultured for 4 days and fixed in 4% PFA in PBS for 30 min.\\n* Cells were washed twice with PBS and incubated with fluorophore-conjugated antibodies (anti-CD81 or anti-CD44) for 1 h.\\n* Antibody signal was measured by flow cytometry gated on live cells based on forward and side scatter.\\n\\n**PerturbView Signal Quantification**\\n\\n* Cell models were fixed in 4% PFA.\\n* After 30 min, cells were washed with PBST three times.\\n* Plates without PerturbView were reserved at 4°C in PBS + 0.4 U/μl Ribolock while decross-linking, and IVT protocols were performed.\\n* Ethanol permeabilization and reverse transcription were performed within 72 h after fixation.\\n\\n**Frameshift Reporter Assay**\\n\\n* Cas9-expressing A549 cells were first transduced with the frameshift reporter at a high MOI by repeating the transduction processes and selected with 300 μg/ml hygromycin for 1 week.\\n* Five targeting and five nontargeting control sgRNAs were individually cloned into a PerturbView vector, pooled equimolarly, packaged as lentivirus, and transduced into Cas9-expressing A549 cells at a low MOI, followed by puromycin selection.\\n* Cells were plated at 12,000 cells per well in a 96-well plate and fixed with 4% PFA for 30 min the following day.\\n* Epitope staining was conducted as previously described, followed by ISS.\\n\\n**Library Construction for the BMDM Screen**\\n\\n* Four sgRNAs were designed per gene for 23 NF-κB-related genes, 130 metabolic and signaling genes, and 10 olfactory receptor genes, along with 60 different nontargeting sgRNAs.\\n* Spacer sequences were generated by crisprVerse to design barcodes that can be uniquely identified by six and eight cycles of ISS for NTC and BMDM libraries, respectively.\\n* The oligonucleotide pool was synthesized by Twist Bioscience, amplified by dial-out PCR, cloned into a PerturbView vector with Golden Gate cloning, and packaged into a lentivirus library while maintaining >1,000× coverage at every step.\\n\\n**NF-κB Translocation Screens**\\n\\n* BMDMs were isolated from mice expressing Cas9–eGFP and cultured in DMEM high glucose, 10% FBS, 2 mM glutamine, 1% penicillin–streptomycin, and 100 ng/ml colony-stimulating factor 1.\\n* After 2 days, these cells were split into new plates and infected with the guide-bearing lentiviruses at an MOI of <0.2.\\n* The medium was replaced the following day, and selection of the infected BMDMs began with 5 μg/ml puromycin.\\n* Three days after the selection and medium change, cells were plated in a 12-well plate at a density of 276,000 cells per well.\\n* After 24 h, cells were stimulated with 10 ng/ml TNF, 2 ng/ml IL-1β, 100 ng/ml LPS, and water for 40 min before fixation.\\n* Cells were fixed in 4% PFA in PBS for 30 min.\\n* Following three PBST washes, cells were permeabilized with 0.2% Triton X-100 in PBS for 10 min.\\n* Cells were then blocked with 3% BSA in PBS for 45 min, incubated with anti-p65 in PBS at 1:1,000 for 1 h, and washed three times with PBST.\\n* The medium was replaced with donkey anti-rabbit CF430 IgG diluted 1:500 in 3% BSA in PBS and incubated for 30 min, followed by three washes with PBST.\\n* Before imaging, the medium was replaced with 200 ng/ml DAPI in 2× saline-sodium citrate buffer for 10 min.\\n* After antibody imaging, ISS was performed beginning from the ethanol permeabilization step.\\n\\nHere's a high-level chronological summary of all the steps:\\n\\n1. Cell culture and maintenance (various cell lines)\\n2. Lentiviral transduction (HEK 293T cells)\\n3. ISS (In Situ Sequencing) protocol development and optimization\\n4. Assessment of editing efficiency (targeting sgRNA transduction and flow cytometry)\\n5. PerturbView signal quantification (cell fixation, decross-linking, and IVT)\\n6. Frameshift reporter assay (Cas9-expressing A549 cells and PerturbView vector)\\n7. Library construction for the BMDM screen (oligonucleotide pool synthesis, amplification, and cloning)\\n8. NF-κB translocation screens (BMDM isolation, lentiviral infection, and ISS)\\n\\nWow, that's a lot of steps! I hope this summary helps you understand the complex protocol.\"},\n",
       " 'flip': True}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Train Personal Reward Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rm = SynthesizeMe(model_id='together_ai/meta-llama/Llama-4-Maverick-17B-128E-Instruct-FP8', num_workers=64)\n",
    "rm = SynthesizeMe(model_id='gemini/gemini-2.0-flash', num_workers=64)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rm.fit(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The synthesized persona is that of a detail-oriented and culturally sensitive individual who values accuracy and flexibility. This person is likely interested in learning about different cultures and languages, particularly names and their meanings. They appreciate thorough explanations and a range of options to choose from, and they prefer solutions that can be adapted to their specific needs. They are also likely to be tech-savvy and appreciate code that is well-written and easy to use. While they may enjoy a creative or poetic style of communication, they prioritize accuracy and usefulness above all else.'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rm.get_persona()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'conversation': '[user]: What would be the english name for name Muzila Nchuminya (of Tonga origin)?\\n\\n[assistant]: After conducting research, I found that the name Muzila Nchuminya is of Tongan origin. In Tongan culture, names often have specific meanings and are composed of a combination of words that convey a particular message or characteristic.\\n\\nWhile there may not be a direct English translation of the name Muzila Nchuminya, I can provide some insights into the possible meanings and components of the name.\\n\\n\"Muzila\" is a common Tongan name that means \"to be strong\" or \"to be powerful.\" It is often given to children born into families of high social status or to those who are expected to excel in their chosen fields.\\n\\n\"Nchuminya\" is a less common name in Tongan culture, but it is believed to mean \"to be wise\" or \"to be intelligent.\" It is often associated with individuals who possess great knowledge, wisdom, and insight.\\n\\nTaking these meanings into account, a possible English translation of the name Muzila Nchuminya could be \"Strong and Wise\" or \"Powerful and Intelligent.\" However, it\\'s essential to note that names can have multiple meanings and interpretations, and the true significance of the name Muzila Nchuminya may be unique to the individual or their family.\\n\\nIn terms of a more direct English name, it\\'s challenging to find a perfect equivalent. However, some possible options that capture the essence of the name Muzila Nchuminya could be:\\n\\n* Maximus (meaning \"greatest\" or \"strongest\")\\n* Vigilantius (meaning \"watchful\" or \"wise\")\\n* Fortunatus (meaning \"fortunate\" or \"powerful\")\\n* Intellectus (meaning \"intelligence\" or \"wisdom\")\\n\\nPlease keep in mind that these suggestions are purely speculative and may not accurately reflect the intended meaning of the name Muzila Nchuminya.\\n\\n[user]: What does the name WIlliam Mutiti mean and what would be a good English alternative?\\n\\n[assistant]: The name William Mutiti is a fascinating combination of cultures. William is a classic English name that means \"resolute protector\" or \"strong-willed warrior.\" It\\'s a name that has been associated with strength, courage, and leadership throughout history.\\n\\nMutiti, on the other hand, is a name of Shona origin, which is a Bantu language spoken in Zimbabwe. In Shona culture, the name Mutiti is believed to mean \"to be brave\" or \"to be fearless.\" It\\'s a name that is often associated with individuals who possess great courage, resilience, and determination.\\n\\nConsidering the meanings and origins of both names, a possible English translation of the name William Mutiti could be \"Brave and Resolute Protector\" or \"Fearless Warrior.\"\\n\\nAs for an English alternative, some possible options that capture the essence of the name William Mutiti could be:\\n\\n* Maximilian (meaning \"greatest\" or \"bravest\")\\n* Atticus (meaning \"bold\" or \"courageous\")\\n* Cassius (meaning \"hollow\" or \"brave\")\\n* Thane (meaning \"warrior\" or \"noble\")\\n\\nHowever, if you\\'re looking for a name that is more directly inspired by the Shona culture, you might consider names like:\\n\\n* Mufaro (meaning \"brave\" or \"fearless\")\\n* Mutasa (meaning \"brave\" or \"strong\")\\n* Chido (meaning \"brilliant\" or \"courageous\")\\n* Tafadzwa (meaning \"brilliant\" or \"excellent\")\\n\\nUltimately, the choice of name will depend on your personal preferences and the cultural context in which you\\'re operating.\\n\\n[user]: What does the name Ohdiambo mean and what would be a good english alternative?\\n\\n[assistant]: The name Ohdiambo is of Luo origin, which is a Bantu language spoken in Kenya and Uganda. In Luo culture, the name Ohdiambo is believed to be a patronymic name, meaning that it is derived from the name of a father or ancestor.\\n\\nThe name Ohdiambo is thought to be a combination of two words: \"Oho\" and \"Diambo.\" \"Oho\" is a Luo word that means \"son of\" or \"descendant of,\" while \"Diambo\" is a name that is believed to mean \"strong\" or \"powerful.\"\\n\\nTaking these meanings into account, a possible English translation of the name Ohdiambo could be \"Son of the Strong\" or \"Descendant of Power.\"\\n\\nAs for an English alternative, some possible options that capture the essence of the name Ohdiambo could be:\\n\\n* Strongson (a direct translation of the name\\'s meaning)\\n* Powerhouse (a name that conveys strength and energy)\\n* Dominick (a name that means \"of the lord\" or \"strong ruler\")\\n* Maximus (a name that means \"greatest\" or \"strongest\")\\n\\nHowever, if you\\'re looking for a name that is more directly inspired by the Luo culture, you might consider names like:\\n\\n* Odhiambo (a variant of the original name, with a slightly different spelling)\\n* Odhiambo Omondi (a name that adds a second name, \"Omondi,\" which means \"gift\" or \"talent\")\\n* Odhiambo Otieno (a name that adds a second name, \"Otieno,\" which means \"born on a Tuesday\" or \"strong and powerful\")\\n* Odhiambo Ochieng (a name that adds a second name, \"Ochieng,\" which means \"born on a Thursday\" or \"strong and courageous\")\\n\\nUltimately, the choice of name will depend on your personal preferences and the cultural context in which you\\'re operating.\\n\\n[user]: what does the name munyambala mean and what would be a good english alternative\\n\\n[assistant]: The name Munyambala is of Tonga origin, which is a Bantu language spoken in Zambia and Zimbabwe. In Tonga culture, the name Munyambala is believed to be a combination of two words: \"Munyam\" and \"Bala.\"\\n\\n\"Munyam\" is a Tonga word that means \"to be strong\" or \"to be powerful.\" It\\'s a name that is often associated with individuals who possess great physical or mental strength.\\n\\n\"Bala\" is a Tonga word that means \"to be wise\" or \"to be intelligent.\" It\\'s a name that is often associated with individuals who possess great knowledge, wisdom, and insight.\\n\\nTaking these meanings into account, a possible English translation of the name Munyambala could be \"Strong and Wise\" or \"Powerful and Intelligent.\"\\n\\nAs for an English alternative, some possible options that capture the essence of the name Munyambala could be:\\n\\n* Maximilian (meaning \"greatest\" or \"strongest\")\\n* Vigilantius (meaning \"watchful\" or \"wise\")\\n* Fortunatus (meaning \"fortunate\" or \"powerful\")\\n* Intellectus (meaning \"intelligence\" or \"wisdom\")\\n\\nHowever, if you\\'re looking for a name that is more directly inspired by the Tonga culture, you might consider names like:\\n\\n* Munyambo (a variant of the original name, with a slightly different spelling)\\n* Munyambwe (a name that adds a suffix \"-we,\" which means \"of\" or \"from\")\\n* Munyambwa (a name that adds a suffix \"-wa,\" which means \"of\" or \"from\")\\n* Munyambwa Mwansa (a name that adds a second name, \"Mwansa,\" which means \"born on a Sunday\" or \"strong and wise\")\\n\\nUltimately, the choice of name will depend on your personal preferences and the cultural context in which you\\'re operating.\\n\\nIt\\'s worth noting that the name Munyambala is also found in other cultures, such as the Lunda people of Angola and the Democratic Republic of Congo, where it is believed to mean \"to be strong and brave\" or \"to be powerful and courageous.\"\\n\\n[user]: What does the name Sally Smith mean and what would be a good tongan alternative?\\n\\n[assistant]: The name Sally Smith is a classic English name that is composed of two separate names: Sally and Smith.\\n\\n\"Sally\" is a shortened form of the name Sarah, which is a Hebrew name that means \"lady\" or \"princess.\" It\\'s a name that has been popular in English-speaking countries for centuries and is often associated with qualities like kindness, compassion, and intelligence.\\n\\n\"Smith\" is a surname that is derived from the Old English word \"smith,\" which means \"blacksmith\" or \"metalworker.\" It\\'s a name that is often associated with qualities like strength, resilience, and craftsmanship.\\n\\nTaking these meanings into account, a possible Tongan translation of the name Sally Smith could be \"Lady of Strength\" or \"Princess of the Forge.\"\\n\\nAs for a Tongan alternative, some possible options that capture the essence of the name Sally Smith could be:\\n\\n* Sione Fale (Sione means \"strong\" or \"powerful,\" and Fale means \"house\" or \"home\")\\n* Sione Finau (Sione means \"strong\" or \"powerful,\" and Finau means \"wise\" or \"intelligent\")\\n* Sione Tavake (Sione means \"strong\" or \"powerful,\" and Tavake means \"brave\" or \"courageous\")\\n* Sione Tupou (Sione means \"strong\" or \"powerful,\" and Tupou means \"to be strong\" or \"to be powerful\")\\n\\nHowever, if you\\'re looking for a name that is more directly inspired by the Tongan culture, you might consider names like:\\n\\n* Sione (a simple and powerful name that means \"strong\" or \"powerful\")\\n* Finau (a name that means \"wise\" or \"intelligent\")\\n* Tavake (a name that means \"brave\" or \"courageous\")\\n* Tupou (a name that means \"to be strong\" or \"to be powerful\")\\n\\nUltimately, the choice of name will depend on your personal preferences and the cultural context in which you\\'re operating.\\n\\n[user]: what does the name Oketch mean and what would be a good english alternative?',\n",
       " 'first_completion': '[assistant]: The name Oketch is of Luo origin, which is a Bantu language spoken in Kenya and Uganda. In Luo culture, the name Oketch is believed to be a patronymic name, meaning that it is derived from the name of a father or ancestor.\\n\\nThe name Oketch is thought to be a combination of two words: \"Oko\" and \"Etch.\" \"Oko\" is a Luo word that means \"son of\" or \"descendant of,\" while \"Etch\" is a name that is believed to mean \"to be wise\" or \"to be intelligent.\"\\n\\nTaking these meanings into account, a possible English translation of the name Oketch could be \"Son of Wisdom\" or \"Descendant of Intelligence.\"\\n\\nAs for an English alternative, some possible options that capture the essence of the name Oketch could be:\\n\\n* Sage (a name that means \"wise\" or \"intelligent\")\\n* Atticus (a name that means \"wise\" or \"courageous\")\\n* Cassius (a name that means \"hollow\" or \"wise\")\\n* Maximus (a name that means \"greatest\" or \"most wise\")\\n\\nHowever, if you\\'re looking for a name that is more directly inspired by the Luo culture, you might consider names like:\\n\\n* Odhiambo (a name that means \"strong\" or \"powerful\")\\n* Omondi (a name that means \"gift\" or \"talent\")\\n* Otieno (a name that means \"born on a Tuesday\" or \"strong and powerful\")\\n* Ochieng (a name that means \"born on a Thursday\" or \"strong and courageous\")\\n\\nAlternatively, you could consider names that reflect the Luo values of wisdom, intelligence, and creativity, such as:\\n\\n* Kipenzi (a name that means \"beloved\" or \"wise\")\\n* Kipkoech (a name that means \"wise\" or \"intelligent\")\\n* Kiprono (a name that means \"wise\" or \"courageous\")\\n* Kipkoech (a name that means \"wise\" or \"intelligent\")\\n\\nUltimately, the choice of name will depend on your personal preferences and the cultural context in which you\\'re operating.',\n",
       " 'second_completion': '[assistant]: The name Oketch is of Luo origin, which is a Bantu language spoken in Kenya and Uganda. In Luo culture, the name Oketch is believed to be a patronymic name, meaning that it is derived from the name of a father or ancestor.\\n\\nThe name Oketch is thought to be a combination of two words: \"Oko\" and \"Etch.\" \"Oko\" is a Luo word that means \"son of\" or \"descendant of,\" while \"Etch\" is a name that is believed to mean \"strong\" or \"powerful.\"\\n\\nTaking these meanings into account, a possible English translation of the name Oketch could be \"Son of Strength\" or \"Descendant of Power.\"\\n\\nAs for an English alternative, some possible options that capture the essence of the name Oketch could be:\\n\\n* Strongson (a direct translation of the name\\'s meaning)\\n* Powerhouse (a name that conveys strength and energy)\\n* Dominick (a name that means \"of the lord\" or \"strong ruler\")\\n* Maximus (a name that means \"greatest\" or \"strongest\")\\n\\nHowever, if you\\'re looking for a name that is more directly inspired by the Luo culture, you might consider names like:\\n\\n* Okello (a variant of the original name, with a slightly different spelling)\\n* Okello Omondi (a name that adds a second name, \"Omondi,\" which means \"gift\" or \"talent\")\\n* Okello Otieno (a name that adds a second name, \"Otieno,\" which means \"born on a Tuesday\" or \"strong and powerful\")\\n* Okello Ochieng (a name that adds a second name, \"Ochieng,\" which means \"born on a Thursday\" or \"strong and courageous\")\\n\\nUltimately, the choice of name will depend on your personal preferences and the cultural context in which you\\'re operating.\\n\\nIt\\'s worth noting that the name Oketch is also found in other cultures, such as the Lunda people of Angola and the Democratic Republic of Congo, where it is believed to mean \"to be strong and brave\" or \"to be powerful and courageous.\"',\n",
       " 'reasoning': 'Both responses are very similar in their structure and content. However, the first response provides a wider range of English alternative names, including some that reflect Luo values of wisdom, intelligence, and creativity. This aligns well with the user\\'s persona, which values thorough explanations and a range of options to choose from. The first response also provides a slightly different interpretation of the name \"Etch,\" which could be seen as more accurate or insightful.',\n",
       " 'preference': 'First'}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rm.get_demos()[0]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gcloud",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "undefined.undefined.undefined"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
